course6/toyapp-adam/math/vector.h

```diff
@@ -1,113 +1,172 @@
 #pragma once
 
+#include <cassert>
+#include <array>
 #include <cmath>
+#include <limits>
 
 namespace Math {
 
-struct Vector2D
+template<typename Type, unsigned Dimensions, typename FloatType = float>
+class Vector
 {
-	float x = 0.0f;
-	float y = 0.0f;
+public:
+	using Component = Type;
 
-	Vector2D operator-() const { return {-x, -y}; }
-	Vector2D operator+(const Vector2D& rhs) const { return {x + rhs.x, y + rhs.y}; }
-	Vector2D operator-(const Vector2D& rhs) const { return {x - rhs.x, y - rhs.y}; }
-	Vector2D operator*(float scalar) const { return {x * scalar, y * scalar}; }
+	Vector()
+	:	mComponents{{0}}
+	{}
 
-	Vector2D& operator+=(const Vector2D& rhs)
+	template<typename... Args>
+	Vector(Args... args)
+	:	mComponents{{args...}}
+	{}
+
+	Vector(const Vector<Type, Dimensions-1>& other, Type lastComponent)
+	{
+		for (unsigned i = 0; i < Dimensions-1; ++i)
+			mComponents[i] = other[i];
+		mComponents[Dimensions-1] = lastComponent;
+	}
+
+	Vector(const Vector& other)
+	{
+		operator=(other);
+	}
+
+	operator Vector<Type, Dimensions-1>() const
+	{
+		Vector<Type, Dimensions-1> result;
+		for (unsigned i = 0; i < Dimensions-1; ++i)
+			result[i] = mComponents[i];
+		return result;
+	}
+
+	friend bool operator==(const Vector& lhs, const Vector& rhs)
```

Comment:
This could be a member function. No need to use ```friend```

Reply:
I get compilation errors when I do that: `C:/lib/boost_1_65_1/boost/test/tools/old/impl.hpp:107:17: error: no match for 'operator==' (operand types are 'const Math::Vector<float, 2>' and 'const Math::Vector<float, 2>')
     return left == right;`

Reply:
was it a const method?